### PR TITLE
YOLOV8Detector [forward pass only]

### PR DESCRIPTION
**Pull request by ianstenbit**
Credit to @leondgarse for the [original port](https://github.com/leondgarse/keras_cv_attention_models/blob/main/keras_cv_attention_models/yolov8/yolov8.py) from Torch.

Colab [demo](https://colab.research.google.com/gist/ianstenbit/91a5b35020630f62cc9fc889ea80ab1d/kerascv-yolov8.ipynb)

Note that this currently _does not_ support training.
After a lot of hacking around with it, I realized I was handling the backwards pass completely wrong.
[This post](https://openmmlab.medium.com/dive-into-yolov8-how-does-this-state-of-the-art-model-work-10f18f74bab1) describes how the loss _should_ work, and the original source for training YOLOv8 is [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/v8/detect/train.py).

I am proposing that we clean up and merge this YOLOv8 model with COCO-trained presets, and we can add a working backwards pass as a next step.

TODO:
- [x] Add tests
- [x] Cleanup anchor generation code
- [x] Add docstrings

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Great work!

**Comment by soma2000-lang** [comment]
@ianstenbit I had a few query to increase my knowledge regarding this matter

1.Is there any particular reason for choosing CSP Darknet as the backbone

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+        outputs = yolov8_head(
+            fpn_features,
+            num_classes,
+            64,  # bbox_len
+            name="head_",
+        )
+        outputs = layers.Activation(
+            "linear", dtype="float32", name="outputs_fp32"
+        )(outputs)
+        super().__init__(inputs=images, outputs=outputs)
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py]: Done (via **kwargs)

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+
+def csp_with_2_conv(
+    inputs,
+    channels=-1,
+    depth=2,
+    shortcut=True,
+    expansion=0.5,
+    activation="swish",
+    name="",
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py]: Done

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    short, deep = tf.split(pre, 2, axis=channel_axis)
+
+    out = [short, deep]
+    for id in range(depth):
+        deep = conv_bn(
+            deep,
+            hidden_channels,
+            kernel_size=3,
+            activation=activation,
+            name=name + "pre_{}_1_".format(id),
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py]: Done throughout.

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+    if kernel_size > 1:
+        inputs = layers.ZeroPadding2D(
+            padding=kernel_size // 2, name=name and name + "pad"
+        )(inputs)
+
+    nn = layers.Conv2D(
+        filters=output_channel,
+        kernel_size=kernel_size,
+        strides=strides,
+        padding="VALID",
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py]: Done

**Comment by ianstenbit** [comment]
> @ianstenbit I had a few query to increase my knowledge regarding this matter
> 
> 1.Is there any particular reason for choosing CSP Darknet as the backbone

CSPDarkNet is the original backbone used for YOLOv5 -- I'm not sure if it's exactly identical to this backbone from YOLOv8, although it looks at least quite similar.

the backbone in this implementation is the same one used by the official implementation

File: keras_cv/models/object_detection/yolo_v8/compat_anchor_generation.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+"""
+I can't get keras_cv.layers.AnchorGenerator to match these numerics, so I've
+copy-pastad this here for now. It needs to go eventually.
```
Comment (jbischof) [inline keras_cv/models/object_detection/yolo_v8/compat_anchor_generation.py:18]: nit: copy-pasta'd

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+)
+from keras_cv.models.task import Task
+from keras_cv.utils.python_utils import classproperty
+from keras_cv.utils.train import get_feature_extractor
+
+BATCH_NORM_EPSILON = 1e-3
+BATCH_NORM_MOMENTUM = 0.97
+
+
+def conv_bn(
```
Comment (jbischof) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:50]: I wonder if we can give this layer more descriptive name like `stack` or `block` a la resnet

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+    strides=1,
+    activation="swish",
+    name=None,
+):
+    if kernel_size > 1:
+        inputs = layers.ZeroPadding2D(
+            padding=kernel_size // 2, name=f"{name}_pad"
+        )(inputs)
+
+    nn = layers.Conv2D(
```
Comment (jbischof) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:63]: s/nn/x

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+        out,
+        input_channels,
+        kernel_size=1,
+        activation=activation,
+        name=f"{name}_output",
+    )
+    return out
+
+
+# This should probably just be a CSPDarknet. That's an outstanding TODO
```
Comment (jbischof) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:166]: Ah in that case my nits on the code above are less relevant

Comment (jbischof) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py:166]: Do you think we could load their weights into our CSPDarknetBackbone? Or are the architectures not canonical?

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+        """Dictionary of preset names and configurations."""
+        return copy.deepcopy(yolo_v8_backbone_presets)
+
+    @classproperty
+    def presets_with_weights(cls):
+        """Dictionary of preset names and configurations that include weights."""
+        return copy.deepcopy(yolo_v8_backbone_presets_with_weights)
+
+
+def path_aggregation_fpn(features, depth=3, name=None):
```
Comment (jbischof) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:250]: I thought YOLO only uses the last layer? Or are all YOLOs not alike?

**Review by jbischof — COMMENTED** [review]
This is really impressive, great start Ian!

I started adding style comments but it seems like we might be able to swap out our own backbones anyway.

**Comment by leondgarse** [comment]
I'm glad you are almost done, even some training code! I'm currently uploading other model weights.
- The `get_anchors` function for yolov8 is actually rather simple. The `base_anchors` calculation there is for `efficientdet`. May refer the official implementation [Github ultralytics/ultralytics/tal.py make_anchors](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/utils/tal.py#L192).
- I modified the pretrained model weights outputing bbox in format `[top, left, bottom right]`, while it's originally `[left, top, right, bottom]`. You may decide which one to use.
- This is the script I used for porting PyTorch weights:
  ```py
  sys.path.append('../ultralytics/')
  import torch
  # from ultralytics import YOLO
  # model = YOLO('yolov8n.pt')  # load an official model
  tt = torch.load('yolov8n.pt')
  _ = tt['model'].eval()
  ss = tt['model'].state_dict()

  from keras_cv_attention_models.yolov8 import yolov8
  mm = yolov8.YOLOV8_N(pretrained=None)

  headers = [
      'head_1_reg_1_conv', 'head_1_reg_1_bn', 'head_1_reg_2_conv', 'head_1_reg_2_bn', 'head_1_reg_3_conv',
      'head_2_reg_1_conv', 'head_2_reg_1_bn', 'head_2_reg_2_conv', 'head_2_reg_2_bn', 'head_2_reg_3_conv',
      'head_3_reg_1_conv', 'head_3_reg_1_bn', 'head_3_reg_2_conv', 'head_3_reg_2_bn', 'head_3_reg_3_conv',
      'head_1_cls_1_conv', 'head_1_cls_1_bn', 'head_1_cls_2_conv', 'head_1_cls_2_bn', 'head_1_cls_3_conv',
      'head_2_cls_1_conv', 'head_2_cls_1_bn', 'head_2_cls_2_conv', 'head_2_cls_2_bn', 'head_2_cls_3_conv',
      'head_3_cls_1_conv', 'head_3_cls_1_bn', 'head_3_cls_2_conv', 'head_3_cls_2_bn', 'head_3_cls_3_conv',
  ]
  specific_match_func = lambda tt: tt[:- len(headers)] + headers

  tail_align_dict = {"output_conv": "pre_0_1_conv", "output_bn": "pre_0_1_bn"}

  from keras_cv_attention_models import download_and_load
  # ss = {}
  download_and_load.keras_reload_from_torch_model(
      ss,
      mm,
      tail_align_dict=tail_align_dict,
      specific_match_func=specific_match_func,
      save_name=mm.name + "_coco.h5",
      do_predict=False,
      do_convert=True,
  )
  ```
  **Convert bboxes output `[left, top, right, bottom]` -> `top, left, bottom, right`**
  ```py
  from keras_cv_attention_models.yolov8 import yolov8
  mm = yolov8.YOLOV8_N(pretrained="yolov8_n_coco.h5")
  heads = [ii.name for ii in mm.layers if ii.name.startswith('head')]
  for ii in range(1, 100):
      layer_name = 'head_{}_reg_3_conv'.format(ii)
      print(f">>>> {layer_name = }")
      if layer_name not in heads:
          break
      conv_layer = mm.get_layer(layer_name)
      new_ww = []
      for ww in conv_layer.get_weights():
          ww = np.reshape(ww, [*ww.shape[:-1], 4, 16])[..., [1, 0, 3, 2], :]
          ww = np.reshape(ww, [*ww.shape[:-2], -1])
          new_ww.append(ww)
      conv_layer.set_weights(new_ww)

  mm.save(mm.name + "_coco.h5")

  from keras_cv_attention_models import test_images, coco
  imm = test_images.dog_cat()
  preds = mm(mm.preprocess_input(imm))
  bboxes, labels, confidences = mm.decode_predictions(preds)[0]
  print(f"{bboxes = }, {labels = }, {confidences = }")
  coco.show_image_with_bboxes(imm, bboxes, labels, confidences, num_classes=80)
  ```

**Comment by ianstenbit** [comment]
@leondgarse -- awesome, thank you for the info! 

I'm going to try to clean up the anchor generation a bit -- I appreciate the pointers! Ideally I can use our `keras_cv.layers.AnchorGenerator` for it :)

The `[top, left, bottom, right]` format is good for us -- we call that `yxyx`, and in KerasCV this one is `rel_yxyx` specifically, since the predictions are relative to the size of the image.

The training code isn't quite working yet. I think I've got some sort of format issue with my target boxes. TBD. I don't really know a lot about how training this model works.

Would you be interested in contributing some of your other models to KerasCV (e.g. YOLOv7, YOLOr, EfficientDet)? You have a very impressive corpus of models in your repo, and we'd love to have you contribute some here if you're interested!

Thank you for your excellent contributions to open-source CV!

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py (hunk 8 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py:166]: I think we probably need some modifications to our CSPDarknetBackbone to support this (namely, right now ours seems to always have 5 stacks? and many of these have only 4).

I'm not certain. It could be that it's not really the same architecture exactly

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py (hunk 9 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py:250]: YOLOv8 does use multiple outputs from the backbone in an FPN, yes.

I think maybe some of the YOLO models don't? Probably @LukeWood or @leondgarse would know for sure

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+
+        # Boxes are now encoded in center_yxhw
+        boxes, classes = self.label_encoder(x, y)
+
+        # Since we're currently using the RetinaNetLabelEncoder, we have
+        # boxes in center_yxhw, and we want bounding_box_format.
+        boxes = bounding_box.convert_format(
+            boxes,
+            source="center_yxhw",
+            target=self.bounding_box_format,
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:497]: why do we want bounding box format?

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+
+        self._has_user_metrics = metrics is not None and len(metrics) != 0
+        self._user_metrics = metrics
+        super().compile(loss=losses, **kwargs)
+
+    def train_step(self, data):
+        x, y = unpack_input(data)
+
+        # Boxes are now encoded in center_yxhw
+        boxes, classes = self.label_encoder(x, y)
```
Comment (LukeWood) [inline keras_cv/models/object_detection/yolo_v8/yolo_v8.py:490]: btw, boxes are also in delta format

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py (hunk 9 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py:250]: > I thought YOLO only uses the last layer? Or are all YOLOs not alike?

I believe the latter is correct - that not all YOLOs are alike.  I believe only the first yolo (maybe first 2-3) only used the final feature map.

Comment (jbischof) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py:48]: Are these YOLOs still less computationally intensive / mobile-friendly than other models in that case? Or do they have as much in common as King Charles II and Charles III?

**Comment by leondgarse** [comment]
- All my detector and classifier model weights are uploaded here in [keras_cv_attention_models/yolov8](https://github.com/leondgarse/keras_cv_attention_models/tree/main/keras_cv_attention_models/yolov8). YOLOV8_N val mAP is tested matching the original result `37.3`. `YOLOV8_X6` is under working.
- For model training, I have been struggling since `yolox` and `yolor`, but the results not satisfying... Gonna try it with yolov8, finding out what are still missing.

**Comment by innat** [comment]
@leondgarse 
It's really interesting there are [classificaiton-mode](https://github.com/leondgarse/keras_cv_attention_models/tree/main/keras_cv_attention_models/yolov8#classification-models) yolo-v8. Have you checked segmentaiton and pose est.? (https://docs.ultralytics.com/tasks/)

**Comment by leondgarse** [comment]
Nop, just because I'm not familiar with those 2 missions...

**Comment by ianstenbit** [comment]
> * For model training, I have been struggling since `yolox` and `yolor`, but the results not satisfying... Gonna try it with yolov8, finding out what are still missing.

Sounds good -- I'm hoping to get training working well here, and I can keep you updated. I've now reverse-engineered the label encoding logic based on the decoding step, and I'm seeing how training works now.

I'm a bit curious about the fact that the final box predictor Conv2D layer has 64 filters, and a weighted sum of these is reduced to the 4 box predictions. I wondered if there is some custom loss that operates on the 64-channel output but I wasn't able to find anything like this in the original implementation. Maybe this is to avoid gradient vanishing?

Anyhow, I'll post some results here if I get anything decent after updating the encoding logic.

**Comment by ianstenbit** [comment]
It turns out I was comically off-base in how I thought the backwards pass worked, and all the work there was for naught. (See the updated PR description with info about the backwards pass)

However, we do have a working forward pass, and this could be a great piece of scaffolding for either:
- someone else to pick this up and try to get training working
- me to come back to this later

File: keras_cv/models/object_detection/yolo_v8/yolo_v8.py (hunk 6 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/yolo_v8/yolo_v8.py]: Done -- I've renamed these to use the backbone naming convention of `apply_xyz`

File: keras_cv/models/__init__.py [hunk 12]
```diff
@@ -96,6 +96,12 @@
 from keras_cv.models.mobilenet_v3 import MobileNetV3Small
 from keras_cv.models.object_detection.faster_rcnn.faster_rcnn import FasterRCNN
 from keras_cv.models.object_detection.retina_net.retina_net import RetinaNet
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_backbone import (
+    YOLOV8Backbone,
+)
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_detector import (
+    YOLOV8Detector,
```
Comment (LukeWood) [inline keras_cv/models/__init__.py:103]: What do we think about YOLOV8BoxDetector?  Detector is basically just adding noise - is it detecting classes?  segmentation maps?  Boxes?

Know BoxDetector is long but I feel this is very undescriptive currently.

**Comment by LukeWood** [comment]
@ianstenbit not sure why, but while the detections look good COCO metrics are bad.

https://colab.research.google.com/drive/1ZwBXxkK3Da_ZC88uKr_184pqi_epOlSL?usp=sharing

Perhaps a bug in my script?

**Comment by LukeWood** [comment]
Measuring the PascalVOC MaP (by ignoring classes) with the COCO weights: not bad!  https://colab.research.google.com/drive/1ZwBXxkK3Da_ZC88uKr_184pqi_epOlSL?usp=sharing

**Comment by ianstenbit** [comment]
I'm going to close this PR and instead re-open one that also includes training (for ease of review).

ETA later today.
